src/marqo/core/document/add_documents_handler.py

```diff
@@ -0,0 +1,562 @@
+            # TODO change the logic when we need to report duplicates as error in the response
+            return
+
+        if not loc and doc_id and doc_id in self.marqo_doc_loc_map:
+            loc = self.marqo_doc_loc_map[doc_id]
+
+        if doc_id in self.marqo_docs:
+            doc_id = self.marqo_docs.pop(doc_id)[ORIGINAL_ID]
+
+        # Even if the last document is invalid, we should not use previous ones?
+        if doc_id:
+            self.visited_doc_ids.add(doc_id)
+
+        self.responses.append((loc, MarqoAddDocumentsItem(
+            id=doc_id if doc_id is not None else '',
+            error=error.error_message,
+            message=error.error_message,
+            status=error.status_code,
+            code=error.error_code
+        )))
+
+        self.errors = True
+
+    def collect_successful_response(self, doc_id: Optional[str]):
+        loc = self.marqo_doc_loc_map.get(doc_id, None)
+
+        self.responses.append((loc, MarqoAddDocumentsItem(
+            id=doc_id if doc_id is not None else '',
+            status=200,
+        )))
+
+    def to_add_doc_responses(self, index_name: str) -> MarqoAddDocumentsResponse:
+        processing_time = (timer() - self.start_time) * 1000
+        # since we reversed the doc list to skip duplicate docs, we now need to reverse the response
+        sorted_responses = [response for _, response in sorted(self.responses, key=lambda r: r[0] or 0, reverse=True)]
+        return MarqoAddDocumentsResponse(errors=self.errors, index_name=index_name, items=sorted_responses,
+                                         processingTimeMs=processing_time)
+
+
+class AddDocumentsHandler(ABC):
+
+    def __init__(self, marqo_index: MarqoIndex, config: Config, add_docs_params: AddDocsParams):
+        self.marqo_index = marqo_index
+        self.add_docs_params = add_docs_params
+        self.config = config
+        self.add_docs_response_collector = AddDocumentsResponseCollector()
+        self.tensor_fields_container = self._create_tensor_fields_container()
+
+    def add_documents(self):
+        """
+        Template method for adding documents to Marqo index
+        """
+        with RequestMetricsStore.for_request().time("add_documents.processing_before_vespa"):
+            for loc, original_doc in enumerate(reversed(self.add_docs_params.docs)):
+                doc = copy.deepcopy(original_doc)
+                original_id = None
+                try:
+                    self.validate_doc(doc)
+
+                    original_id = self.validate_and_pop_doc_id(doc)
+                    doc_id = original_id or str(uuid.uuid4())
+                    marqo_doc = {ORIGINAL_ID: original_id, MARQO_DOC_ID: doc_id}  # keep this info for error report
+
+                    for field_name, field_content in doc.items():
+                        self.handle_field(marqo_doc, field_name, field_content)
+
+                    self.handle_multi_modal_fields(marqo_doc)
```

Comment:
I think these methods should have a _ prefix as they're not to be accessed outside of the class

Reply:
I'm not too sure about python convention for this. usually underscore prefixed method means private method or protected one. Since these method are only used in the class hierarchy and not by other classes. I think it makes sense to have a `_` prefix. I'll make the change